### PR TITLE
Salesforce Sync Refactoring

### PR DESCRIPTION
**Pull request by alanhartless**
[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | y
| New feature? | 
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
The Salesforce sync became complicated and riddled with issues. This PR refactors the pushLeads method to be able to write tests and ensure all is working accordingly.

This PR includes the following PRs:
#4066

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. 
2. 

#### Steps to test this PR:
1. 
2. 

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.

### CODE REVIEW

**Comment by escopecz** [comment]
I'm getting this notice. All I did was apply this PR, clear the cache and run the fetch command
```
$ php app/console mautic:integration:fetchleads --integration=Salesforce
Command to fetch contacts from integration
Fetch contacts command is starting
0 contacts were updated and 0 contacts were created

Fetching companies
Fetch companies command is starting
0 companies were updated and 0 companies were created

Updating/creating leads from Mautic to Salesforce
About 2 to update and about 6 to create/update
 1/8 [===>------------------------]  12%PHP Notice:  Undefined index: integration_entity in plugins/MauticCrmBundle/Integration/SalesforceIntegration.php on line 2104
 8/8 [============================] 100%
Number of contacts pushed: 0 updated, 1 created, 2 had errors and 4 were ignored (likely duplicates)
```

**Comment by escopecz** [comment]
I fixed the PHP notice. Not sure if there isn't a problem before the fix which should have a deeper consequences, but the fix won't harm anything.

But the SF lead does not update. Here is what I did:
1. Create a Mautic contact with firstname = John's, lastname = SFtest, email = sftest@john.email
2. Run the fetch command - the contact appears at SF as a new lead
3. Update the Mautic contact like firstname = John's2, lastname = SFtest2
4. Run the fetch command - the SF lead is not updated with new first and last name

**Comment by escopecz** [comment]
I applied the latest changes and updating a SF lead based on contact updates in Mautic contact works. Thanks @mqueme!

But if I convert the SF lead to SF contact, then update the Mautic contact and run the fetch command, it does not update the SF contact. It should, right?

**Comment by isleshocky77** [comment]
The PR is still using `$item['body'][0]['message']` raw. 
Please see #4084 for issues with doing this.

**Comment by escopecz** [comment]
I noticed 2 issues:
1. The company "Unknown" was create in my Mautic and the contact which was created in Mautic and sent over to SF got this company assigned.
2. When I converted SF lead which was created from Mautic to SF contact and run the command, this happened:
```
$ php app/console mautic:integration:fetchleads --integration=Salesforce --time-interval="1 minute"
Command to fetch contacts from integration
Fetch contacts command is starting
0 contacts were updated and 0 contacts were created

Fetching companies
Fetch companies command is starting
0 companies were updated and 0 companies were created

Updating/creating leads from Mautic to Salesforce
About 1 to update and about 1 to create/update
PHP Notice:  Undefined index: ConvertedContactId in plugins/MauticCrmBundle/Integration/SalesforceIntegration.php on line 2156
 1/2 [==============>-------------]  50%

  [Doctrine\ORM\EntityNotFoundException]
  Entity of type 'Mautic\PluginBundle\Entity\IntegrationEntity' for IDs id(19764) was not found
```

**Comment by mqueme** [comment]
This now works for me. Thank you for the hard work!

**Comment by isleshocky77** [comment]
@alanhartless I'm having an issue still that I was having before the refactor. One of the fields on Salesforce is a formula.  Contact.Field1__c which is a formula "Account(Field1__c)".  When the sync runs I get the following error

```Salesforce Error
Unable to create/update fields: Field1__c. Please check the security settings of this field and verify that it is read/write for your profile or permission set. Contact ID# XXXXXXXXX
````

I changed the sync directory to only pull, but it looks like it is still attempting to push to salesforce.com.

This doesn't happen on the Salesforce Lead object as there, the Field is a text field.

**Comment by isleshocky77** [comment]
@alanhartless That looks to have not had any effect. I'm still getting the same error after pulling.

**Comment by isleshocky77** [comment]
@alanhartless Sync now bombs

```
sostrow@islanders /srv/mautic/project $ ./app/console mautic:integration:synccontacts -vvv -i Salesforce -a '1 days'
Command to fetch contacts from integration
Fetch contacts command is starting
18 contacts were updated and 0 contacts were created

Updating/creating leads from Mautic to Salesforce
About 17 to update and about 0 to create/update
 17/17 [============================] 100%
                                                                                                                                                        
  [Doctrine\ORM\EntityNotFoundException]                                                                                                                
  Entity of type 'Mautic\PluginBundle\Entity\IntegrationEntity' for IDs id(Mautic\PluginBundle\Entity\IntegrationEntity with ID #270758) was not found  
                                                                                                                                                        

Exception trace:
 () at /srv/mautic/project/vendor/doctrine/orm/lib/Doctrine/ORM/EntityNotFoundException.php:47
 Doctrine\ORM\EntityNotFoundException::fromClassNameAndIdentifier() at /srv/mautic/project/vendor/doctrine/orm/lib/Doctrine/ORM/Proxy/ProxyFactory.php:195
 Doctrine\ORM\Proxy\ProxyFactory->Doctrine\ORM\Proxy\{closure}() at /srv/mautic/project/app/cache/prod/doctrine/orm/Proxies/__CG__MauticPluginBundleEntityIntegrationEntity.php:296
 Closure->__invoke() at /srv/mautic/project/app/cache/prod/doctrine/orm/Proxies/__CG__MauticPluginBundleEntityIntegrationEntity.php:296
 Proxies\__CG__\Mautic\PluginBundle\Entity\IntegrationEntity->setLastSyncDate() at /srv/mautic/project/plugins/MauticCrmBundle/Integration/SalesforceIntegration.php:1929
 MauticPlugin\MauticCrmBundle\Integration\SalesforceIntegration->processCompositeResponse() at /srv/mautic/project/plugins/MauticCrmBundle/Integration/SalesforceIntegration.php:2345
 MauticPlugin\MauticCrmBundle\Integration\SalesforceIntegration->makeCompositeRequest() at /srv/mautic/project/plugins/MauticCrmBundle/Integration/SalesforceIntegration.php:1134
 MauticPlugin\MauticCrmBundle\Integration\SalesforceIntegration->pushLeads() at /srv/mautic/project/app/bundles/PluginBundle/Command/FetchLeadsCommand.php:201
 Mautic\PluginBundle\Command\FetchLeadsCommand->execute() at /srv/mautic/project/vendor/symfony/console/Command/Command.php:257
 Symfony\Component\Console\Command\Command->run() at /srv/mautic/project/vendor/symfony/console/Application.php:872
 Symfony\Component\Console\Application->doRunCommand() at /srv/mautic/project/vendor/symfony/console/Application.php:193
 Symfony\Component\Console\Application->doRun() at /srv/mautic/project/vendor/symfony/framework-bundle/Console/Application.php:92
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /srv/mautic/project/vendor/symfony/console/Application.php:124
 Symfony\Component\Console\Application->run() at /srv/mautic/project/app/console:41

mautic:integration:fetchleads [-i|--integration INTEGRATION] [-d|--start-date START-DATE] [-t|--end-date END-DATE] [-a|--time-interval [TIME-INTERVAL]] [-l|--limit [LIMIT]] [-f|--force] [-h|--help] [-q|--quiet] [-v|vv|vvv|--verbose] [-V|--version] [--ansi] [--no-ansi] [-n|--no-interaction] [-s|--shell] [--process-isolation] [-e|--env ENV] [--no-debug] [--] <command>

sostrow@islanders /srv/mautic/project $ 
```

**Comment by isleshocky77** [comment]
@alanhartless The Update error still appears to be there as well in Notifications.

**Comment by isleshocky77** [comment]
With the latest code base I got the following status output which doesn't make sense to me

```Updating/creating leads from Mautic to Salesforce
About 6 to update and about 0 to create/update
 6/6 [============================] 100%
Number of contacts pushed: 0 updated, 0 created, 0 had errors and 0 were ignored (likely duplicates)
```

File: plugins/MauticCrmBundle/Api/SalesforceApi.php [hunk 1]
```diff
@@ -267,7 +310,7 @@ public function getLeads($query, $object)
 
             $getLeadsQuery = 'SELECT '.$fields.' from '.$object.' where LastModifiedDate>='.$query['start'].' and LastModifiedDate<='.$query['end'].$ignoreConvertedLeads;
```
Comment (isleshocky77) [inline plugins/MauticCrmBundle/Api/SalesforceApi.php]: Need to force the format of the start and end dates.  It has to be ISO-8601. 

Example, the following will find no results even if there are.

<pre>
# Returns nothing
./app/console -vvv mautic:integration:synccontacts -i Salesforce -d '2016-07-01' -t '2016-08-01'
</pre>

versus this actually working

<pre>
# Return 500+ results
./app/console -vvv mautic:integration:synccontacts -i Salesforce -d '2016-07-01T00:00:00Z' -t '2016-08-01T00:00:00Z'
</pre>

**Comment by isleshocky77** [comment]
It doesn't appear the **Contacts** (Salesforce Contact) are being associated with their Respective **Company** (Salesforce Account) when being pulled down.

**Comment by mqueme** [comment]
This PR works for me with all the latest changes.

**Comment by escopecz** [comment]
It doesn't work for me. Here is what I do:
1. Create the contact in Mautic
2. Run the fetch command
3. The SF lead is created
4. I update the Mautic contact last name
5. Run the fetch command
6. The SF lead is not updated, Mautic contact is reseted back to the previous state
The arrow for last name field is from Mautic to SF

**Comment by escopecz** [comment]
This happened to me:
```
Updating/creating leads from Mautic to Salesforce
About 1 to update and about 0 to create/update
 1/1 [============================] 100%

  [Doctrine\ORM\EntityNotFoundException]
  Entity of type 'Mautic\PluginBundle\Entity\IntegrationEntity' for IDs id(Mautic\PluginBundle\Entity\IntegrationEntity with ID #22220) was not found
```
Here are steps I took:
1. Create a contact in Mautic
2. Run the fetch command - the contact will appear as SF lead
3. Convert the lead to Contact
4. Assign the SF contact a new company
5. Run the fetch command

**Comment by alanhartless** [comment]
Actually @escopecz, I can't seem to replicate this. Are you able to replicate it?

**Comment by matishaladiwala** [comment]
company and campaign update using a campaign action works!

**Pull request merged**